### DISCUSSION TITLE
OpenID Connect jwt claim 'azp' is mandatory not optional

### DISCUSSION THREAD

**Discussion by matf3en** [nextauthjs/next-auth#6966] [Help] [answered]
### Environment

System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    Memory: 451.27 MB / 7.75 GB
  Binaries:
    Node: 16.14.2 - C:\Program Files\nodejs\node.EXE
    npm: 8.5.5 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (110.0.1587.69)
    Internet Explorer: 11.0.19041.1566

### Reproduction URL

private

### Describe the issue

A provider (Oracle Netsuite) is returning a jwt that does not contain the 'azp' claim. 

Reading the RFCs etc, the azp claim is *optional*.

Next-Auth however checks for the presence of the azp claim and fails if none is found.

The error produced is 

`https://next-auth.js.org/errors#oauth_callback_error missing required JWT property azp {
  error: RPError: missing required JWT property azp`

So this bug is : that Next-Auth (or OpenId-Client) makes an optional claim into a required claim, which it should not.

### How to reproduce

Connect via a standard OAuth2 OIDC flow with Oracle Netsuite from an external app. Observe the error

https://next-auth.js.org/errors#oauth_callback_error missing required JWT property azp {
  error: RPError: missing required JWT property azp

### Expected behavior

If the azp claim was optional not required, this error would not occur.

**Comment by balazsorban44** [comment]
Hard to say much without a reproduction or any more information. Usually `azp` defaults to `client_id`, but in your case the AP might be sending multiple `aud` values, causing the check to be activated for a valid `azp` value. See the code for the underlying library

https://github.com/panva/node-openid-client/blob/ae1222c60541a87c650ce9bff8cc3ab5af9b413d/lib/client.js#L968

In any case, this is not a bug, likely a misconfiguration or a non-compliant provider. (Haven't used Oracle Netsuite so I cannot confirm)

**Reply by jdpnielsen** [reply]
Is there any way to pass `additionalAuthorizedParties` from the provider configuration to the open-id [client](https://github.com/panva/node-openid-client/blob/main/docs/README.md#new-clientmetadata-jwks-options)? The field is required by the open-id client to validate the jwt, if there are multiple `aud`'s & `azp` is set.

**Reply by progeriof** [reply]
+1 for this, I'm currently trying to sign in using id_token on next_auth with an external app, and I need to add this other google_id to this `additionalAuthorizedParties` property.
Have you solved it somehow?

**Reply by balazsorban44** [reply, marked as answer]
https://next-auth.js.org/configuration/providers/oauth#client-option

**Reply by jdpnielsen** [reply]
@balazsorban44 This actually does not answer the the question. The `client` option allows configuration of `open-id`'s `metadata`, but not its `options`, which contains the `additionalAuthorizedParties` parameter.
See https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/core/lib/oauth/client.ts#L37 & https://github.com/panva/node-openid-client/blob/main/docs/README.md#new-clientmetadata-jwks-options

@progeriof I solved it by adding a patch using [patch-package](https://www.npmjs.com/package/patch-package): 

```patch
diff --git a/node_modules/next-auth/core/lib/oauth/client.js b/node_modules/next-auth/core/lib/oauth/client.js
index c2f0d40..9074cdf 100644
--- a/node_modules/next-auth/core/lib/oauth/client.js
+++ b/node_modules/next-auth/core/lib/oauth/client.js
@@ -31,7 +31,7 @@ async function openidClient(options) {
     client_secret: provider.clientSecret,
     redirect_uris: [provider.callbackUrl],
     ...provider.client
-  }, provider.jwks);
+  }, provider.jwks, provider.clientOptions);
   client[_openidClient.custom.clock_tolerance] = 10;
   return client;
 }
\ No newline at end of file
diff --git a/node_modules/next-auth/providers/oauth.d.ts b/node_modules/next-auth/providers/oauth.d.ts
index c2adf84..12fd796 100644
--- a/node_modules/next-auth/providers/oauth.d.ts
+++ b/node_modules/next-auth/providers/oauth.d.ts
@@ -86,6 +86,10 @@ export interface OAuthConfig<P> extends CommonProviderOptions, PartialIssuer {
     profile: (profile: P, tokens: TokenSet) => Awaitable<User>;
     checks?: ChecksType | ChecksType[];
     client?: Partial<ClientMetadata>;
+    /**
+     * Allows configuration of additional open-id client options
+     */
+    clientOptions?: Partial<ClientMetadata>;
     jwks?: {
         keys: JWK[];
     };
diff --git a/node_modules/next-auth/src/providers/oauth.ts b/node_modules/next-auth/src/providers/oauth.ts
index e19d654..0980a23 100644
--- a/node_modules/next-auth/src/providers/oauth.ts
+++ b/node_modules/next-auth/src/providers/oauth.ts
@@ -123,6 +123,10 @@ export interface OAuthConfig<P> extends CommonProviderOptions, PartialIssuer {
   profile: (profile: P, tokens: TokenSet) => Awaitable<User>
   checks?: ChecksType | ChecksType[]
   client?: Partial<ClientMetadata>
+	/**
+	 * Allows configuration of additional open-id client options
+	 */
+	clientOptions?: Partial<ClientMetadata>;
   jwks?: { keys: JWK[] }
   clientId?: string
   clientSecret?: string
```

@balazsorban44 If you are open to a PR to remedy this, i'll open it. Do you have suggestions for the parameter naming? I'm thinking it should be something other than `clientOptions` .
